src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/PatternMaskerProvider.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.credentialsbinding.masking;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.TreeSet;
+import java.util.regex.Pattern;
+
+/**
+ * Quotes strings according to quotation rules of various programs. Sometimes confused with escaping, quoting is done
+ * to pass literal strings in a shell or other interpreter.
+ *
+ * @since TODO
+ */
+public interface PatternMaskerProvider extends ExtensionPoint {
+
+    @Nonnull Collection<String> getAlternativeForms(@Nonnull String input);
+
+    static @Nonnull ExtensionList<PatternMaskerProvider> all() {
+        return ExtensionList.lookup(PatternMaskerProvider.class);
+    }
+
+    @Restricted(NoExternalUse.class)
+    static @Nonnull Collection<String> getAllAlternateForms(@Nonnull String input) {
+        Collection<String> alternateForms = new TreeSet<>(BY_REVERSE_LENGTH);
```

Comment:
At this point, a basic `HashSet` is sufficient. For the reader it's harder to understand why you put a tree here and in the next method, while only the latter one is required.

Reply:
Probably an over optimization since the set sorting doesn't need to happen until the regex is created.